clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -60,23 +60,73 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
                 ),
                 applicability,
             );
+            return true;
+        }
+        return utils::match_borrows_parameter(cx, qpath).map_or(false, |span| {
+            let mut applicability = Applicability::MachineApplicable;
+            span_lint_and_sugg(
+                cx,
+                REDUNDANT_ALLOCATION,
+                hir_ty.span,
+                "usage of `Rc<&T>`",
+                "try",
+                snippet_with_applicability(cx, span, "..", &mut applicability).to_string(),
+                applicability,
+            );
             true
-        } else {
-            utils::match_borrows_parameter(cx, qpath).map_or(false, |span| {
-                let mut applicability = Applicability::MachineApplicable;
-                span_lint_and_sugg(
-                    cx,
-                    REDUNDANT_ALLOCATION,
-                    hir_ty.span,
-                    "usage of `Rc<&T>`",
-                    "try",
-                    snippet_with_applicability(cx, span, "..", &mut applicability).to_string(),
-                    applicability,
-                );
-                true
-            })
+        });
+    }
+
+    if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
+        if let Some(ty) = is_ty_param_diagnostic_item(cx, qpath, sym::Arc) {
+            let mut applicability = Applicability::MachineApplicable;
```

Comment:
This is from what I can tell a copy of the `Rc<T>` implementation above with slight adjustments when it comes to the checked type and lint messages. Could you move the code into its own check function that can be reused for both types?

You can use the `format!()` macro to adjust the message for the specific type. Example for formatting the lint message before emission:

```rust
// assuming that `type_sym` is the `Symbol` of the checked type
span_lint_and_sugg(
    cx,
    REDUNDANT_ALLOCATION,
    hir_ty.span,
    &format!("usage of `{}<Box<T>>`", type_sym), // formatted message
    "try",
    format!(
        "{}<{}>",
        type_sym,
        snippet_with_applicability(cx, inner_span, "..", &mut applicability)
    ),
    applicability,
);
```

Reply:
You can resolve this comment :)